hydromt/data_source/data_source.py

```diff
@@ -0,0 +1,166 @@
+
+from hydromt._typing import DataType
+from hydromt.data_adapter.caching import _uri_validator
+from hydromt.data_adapter.data_adapter_base import DataAdapterBase
+from hydromt.driver import BaseDriver
+
+logger: Logger = getLogger(__name__)
+
+T = TypeVar("T")
+
+
+class DataSource(BaseModel, ABC):
+    """
+    A DataSource is a parsed section of a DataCatalog.
+
+    The DataSource, specific for a data type within HydroMT, is responsible for
+    validating the input from the DataCatalog, to
+    ensure the workflow fails as early as possible. A DataSource has information on
+    the driver that the data should be read with, and is responsible for initializing
+    this driver.
+    """
+
+    _used: bool = PrivateAttr(default=False)
+
+    name: str
+    uri: str
+    data_adapter: DataAdapterBase
+    driver: BaseDriver
+    data_type: ClassVar[DataType]
+    root: Optional[str] = Field(default=None)
+    version: Optional[str] = Field(default=None)
+    provider: Optional[str] = Field(default=None)
+    crs: Optional[int] = Field(default=None)
+    extent: Dict[str, Any] = Field(default_factory=dict)
+
+    def summary(self) -> Dict[str, Any]:
+        """Return a summary of the DataSource."""
+        summ: Dict[str, Any] = self.model_dump(include={"uri"})
+        summ.update(
+            {
+                "data_type": self.__class__.data_type,
+                "driver": self.driver.__repr_name__(),
+                **self.data_adapter.meta,
+            }
+        )
+        return summ
+
+    # TODO: def to_file(self, **query_params) https://github.com/Deltares/hydromt/issues/840
+
+    @model_validator(mode="before")
+    @classmethod
+    def _push_down_data_adapter_args(cls, data: Any):
+        """Copy unit_add, unit_mult and rename to metadata_resolver.
+
+        The metadata resolver should query the data with the right variables and
+        the correct time variables, and this needs these arguments when used in
+        conjunction with the DataAdapter. The DataSource is responsable to keep these
+        up-to-date.
+        """
+        if isinstance(data, dict) or isinstance(data, BaseModel):
+            for da_arg in ["unit_add", "unit_mult", "rename"]:
+                value: Any = get_nested_var(["data_adapter", da_arg], data, {})
+                try:
+                    set_nested_var(["driver", "metadata_resolver", da_arg], data, value)
+                except ValueError:
+                    pass  # let pydantic handle errors when set_nested_var cannot set.
+        return data
+
+    @model_validator(mode="before")
+    @classmethod
+    def _validate_data_type(cls, data: Any) -> Any:
+        """Pydantic does not check class variables, so it is checked here."""
+        if (
+            isinstance(data, dict)
+            and data.get("data_type")
+            and data.get("data_type") != cls.data_type
+        ):
+            raise ValueError(f"'data_type' must be '{cls.data_type}'.")
+        return data
+
```

Comment:
Is this still needed? Because it is not longer possible to initialize a RasterDatasetSource from DataSource right?

```py 
source = DataSource(
    name="test",
    uri="test.zarr",
    driver={"name": "zarr"},
    data_type="RasterDataset",
)
```

Reply:
Without this you can still:
```python
source = RasterDataSource(
    name="test",
    uri="test.zarr",
    driver={"name": "zarr"},
    data_type="GeoDataFrame"
)
```

Reply:
Check. But should data_type be an argument of the DataSource subclass? I would expect it to be a class variable.